pytest_xprocess.py

```diff
@@ -53,3 +53,18 @@ def pytest_runtest_makereport(item, call):
             longrepr = getattr(report, "longrepr", None)
             if hasattr(longrepr, "addsection"):  # pragma: no cover
                 longrepr.addsection("%s log" % name, content)
+
+
+def pytest_unconfigure(config):
```

Comment:
instead of using __del__ as magical transfer, i propose making xprocess a context manager and using the session scope xprocess fixture to do the cleanup in a better integrated way

Reply:
Hey, thanks for the tip! Do you mean something like
```python
@pytest.fixture(scope="session")
def xprocess(request):
    """Return session-scoped XProcess helper to manage long-running
    processes required for testing.
    """
    rootdir = getrootdir(request.config)

    with XProcess(request.config, rootdir) as xproc:
        yield xproc

    # clean up here
```
This would indeed look better but we would have problems with `pytest_runtest_makereport` calling `read` on closed file streams during teardown runtest phase. Perhaps I am missing something here?

Reply:
i will have to check where exactly session scope is torn down, i believe it happens after the last normal test teardown

Reply:
I have just confirmed that `pytest_runtest_makereport ` does complain about the closed file stream with an ugly `INTERNALERROR`. Something like the following would work, but I wonder if this is any better than using the magic pass in `__del__`

```python
@pytest.fixture(scope="session")
def xprocess(request):
    """Return session-scoped XProcess helper to manage long-running
    processes required for testing.
    """
    rootdir = getrootdir(request.config)

    with XProcess(request.config, rootdir) as xproc:
        yield xproc

    """Reading processes exit status is a requirement of subprocess module,
    so each process instance should be properly waited upon."""
    if xproc.running_procs:
        for p in xproc.running_procs:
            p.wait(xproc.proc_wait_timeout)

    """_file_handles will be used to close streams in pytest_unconfigure hook""""

    request.config.__dict__.setdefault("_file_handles", xproc.file_handles)

# ...

def pytest_unconfigure(config):
    file_handles = getattr(config, "_file_handles", None)
    if file_handles:
        for f in file_handles:
            f.close()
```
We still use the magic transfer but at least it's done during session teardow from the fixture and not from the XProcess class. I don't know if this is any better than the current approach, though.